cmd/stack_ls.go

```diff
@@ -78,69 +81,112 @@ func newStackLsCmd() *cobra.Command {
 				return stackSummaries[i].Name().String() < stackSummaries[j].Name().String()
 			})
 
-			_, showURLColumn := b.(httpstate.Backend)
-
-			// Devote 48 characters to the name width, unless there is a longer name.
-			maxName := 47
-			for _, summary := range stackSummaries {
-				name := summary.Name().String()
-				if len(name) > maxName {
-					maxName = len(name)
-				}
+			if jsonOut {
+				return formatJSON(b, current, stackSummaries)
 			}
-			maxName++ // Account for adding the '*' to the currently selected stack.
-
-			// Header string and formatting options to align columns.
-			formatDirective := "%-" + strconv.Itoa(maxName) + "s %-24s %-18s"
-			headers := []interface{}{"NAME", "LAST UPDATE", "RESOURCE COUNT"}
-			if showURLColumn {
-				formatDirective += " %s"
-				headers = append(headers, "URL")
-			}
-			formatDirective = formatDirective + "\n"
 
-			fmt.Printf(formatDirective, headers...)
-			for _, summary := range stackSummaries {
-				const none = "n/a"
+			return formatConsole(b, current, stackSummaries)
+		}),
+	}
+	cmd.PersistentFlags().BoolVarP(
+		&jsonOut, "json", "j", false, "Emit outputs as JSON")
+	cmd.PersistentFlags().BoolVarP(
+		&allStacks, "all", "a", false, "List all stacks instead of just stacks for the current project")
 
-				// Name column
-				name := summary.Name().String()
-				if name == current {
-					name += "*"
-				}
+	return cmd
+}
 
-				// Last update column
-				lastUpdate := none
-				if stackLastUpdate := summary.LastUpdate(); stackLastUpdate != nil {
-					lastUpdate = humanize.Time(*stackLastUpdate)
-				}
+func formatJSON(b backend.Backend, currentStack string, stackSummaries []backend.StackSummary) error {
+	type stackSummaryJSON struct {
```

Comment:
We may want to add comments here that we cannot break the interface here.  With it being an internal type (even internal to this function!) it could look like it is okay to change but will be something users take a hard dependency on.

Reply:
Promoted the type to top level (it remains unexported for now) and added a comment.